src/Log/Engine/BaseLog.php

```diff
@@ -90,6 +91,46 @@ public function scopes()
      */
     protected function _format(string $message, array $context = []): string
     {
-        return $message;
+        if (strpos($message, '{') === false && strpos($message, '}') === false) {
+            return $message;
+        }
+
+        preg_match_all('/\{([a-z][a-z0-9-_]*)\}/i', $message, $matches);
+        if (empty($matches)) {
+            return $message;
+        }
+
+        $placeholders = array_intersect($matches[1], array_keys($context));
+        $replacements = [];
+
+        foreach ($placeholders as $key) {
+            $value = $context[$key];
+
+            if (is_scalar($value)) {
+                $replacements['{' . $key . '}'] = (string)$value;
+                continue;
+            }
+
+            if (is_object($value)) {
+                if (method_exists($value, '__toString')) {
+                    $replacements['{' . $key . '}'] = (string)$value;
+                    continue;
+                }
+
+                if ($value instanceof JsonSerializable) {
+                    $replacements['{' . $key . '}'] = json_encode($value, JSON_UNESCAPED_UNICODE);
+                    continue;
+                }
+
+                if (method_exists($value, 'toArray')) {
+                    $replacements['{' . $key . '}'] = print_r($value->toArray(), true);
+                    continue;
+                }
+            }
+
+            $replacements['{' . $key . '}'] = print_r($value, true);
```

Comment:
This `print_r()` causes PHP to run out of memory in tests :slightly_smiling_face:.

Similar statement exists in 3.x too but I guess it doesn't have any code coverage.
https://github.com/cakephp/cakephp/blob/2f68c080ad27ecb9c9ee19e58961497a75703f7e/src/Log/Engine/BaseLog.php#L110

So I guess the best course of action is to simply ignore objects which cannot be serialized to string or json encoded.

Reply:
I'm :+1: for removing this code. If we find a non-scalar value that can't be formatted can we include `[unhandled value of $typeOfThing]` placeholder so the user can figure out why data is missing?